protocols/protocols.go

```diff
@@ -0,0 +1,518 @@
+	var err error
+	dbHandle, err := sql.Open("postgres", psqlInfo)
+	if err != nil {
+		panic(err)
+	}
+
+	err = dbHandle.Ping()
+	if err != nil {
+		panic(err)
+	}
+	return dbHandle
+}
+
+func assertError(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func assertTxnError(err error, txn *sql.Tx) {
+	if err != nil {
+		txn.Rollback()
+		panic(err)
+	}
+}
+
+func (manager *ProtocolManagerT) populateEventTypes() {
+
+	eventTypesSelectSQL := fmt.Sprintf(`SELECT * FROM %s`, EVENT_TYPES_TABLE)
+
+	rows, err := manager.dbHandle.Query(eventTypesSelectSQL)
+	assertError(err)
+	defer rows.Close()
+
+	for rows.Next() {
+		var eventType EventTypeT
+		err := rows.Scan(&eventType.ID, &eventType.UUID, &eventType.WriteKey, &eventType.EvType,
+			&eventType.EventIdentifier, &eventType.CreatedAt)
+
+		assertError(err)
+		manager.updateEventTypeCache(&eventType, false)
+	}
+}
+
+func (manager *ProtocolManagerT) populateSchemaVersions() {
+
+	schemaVersionsSelectSQL := fmt.Sprintf(`SELECT * FROM %s`, SCHEMA_VERSIONS_TABLE)
+
+	rows, err := manager.dbHandle.Query(schemaVersionsSelectSQL)
+	assertError(err)
+	defer rows.Close()
+
+	for rows.Next() {
+		var schemaVersion SchemaVersionT
+		err := rows.Scan(&schemaVersion.ID, &schemaVersion.UUID, &schemaVersion.EventID, &schemaVersion.SchemaHash,
+			&schemaVersion.Schema, &schemaVersion.Metadata, &schemaVersion.FirstSeen, &schemaVersion.LastSeen)
+		assertError(err)
+
+		manager.updateSchemaVersionCache(&schemaVersion, false, false)
+	}
+}
+
+// This should be called during the Setup() to populate existing event Schemas
+func (manager *ProtocolManagerT) populateEventSchemas() {
+	manager.populateEventTypes()
+	manager.populateSchemaVersions()
+}
+
+//TODO: Support for prefix based
+func computeVersion(event EventT) (schemaHash string, schema map[string]string) {
+
+	eventMap := map[string]interface{}(event)
+
+	flattenedEvent, err := flatten.Flatten((eventMap), "", flatten.DotStyle)
+
+	if err != nil {
+		fmt.Println(err)
+		panic("Failed to flatten the event")
+	} else {
+		fmt.Println(flattenedEvent)
+		finalSchema := make(map[string]string)
+		keys := make([]string, 0, len(finalSchema))
+		for k, v := range flattenedEvent {
+			keys = append(keys, k)
+			reflectType := reflect.TypeOf(v)
+			finalSchema[k] = reflectType.String()
+		}
+		fmt.Println(finalSchema)
+		sort.Strings(keys)
+
+		var sb strings.Builder
+		for _, k := range keys {
+			sb.WriteString(k)
+			sb.WriteString(":")
+			sb.WriteString(finalSchema[k])
+			sb.WriteString(",")
+		}
+		return misc.GetMD5Hash(sb.String()), finalSchema
+	}
+}
+
+//TODO: Use Migrations library
+func (manager *ProtocolManagerT) setupTables() {
+	createEventTypesSQL := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+		id SERIAL PRIMARY KEY,
+		uuid VARCHAR(36) NOT NULL,
+		write_key VARCHAR(32) NOT NULL,
+		ev_type TEXT NOT NULL,
+		event_identifier TEXT NOT NULL DEFAULT '',
+		created_at TIMESTAMP NOT NULL DEFAULT NOW()
+	)
+	`, EVENT_TYPES_TABLE)
```

Comment:
names are a bit confusing. there is a event type `track` and also an event type `Track + App Opened` etc. Consider renaming it?

Reply:
Renamed event_types to event_models